### PR TITLE
multi: Add getblockchaininfo rpc.

### LINKED ISSUE #1475: [rpcserver] Implement getblockchaininfo RPC.
In order to provide additional details about the current state of the blockchain, a new RPC named `getblockchaininfo` should be implemented.

The command for this RPC is [already defined](https://github.com/decred/dcrd/blob/979d21f6f3695f0081ccc0c55994ecdff39b5a27/dcrjson/chainsvrcmds.go#L352) along with its [result](https://github.com/decred/dcrd/blob/master/dcrjson/chainsvrresults.go#L94-L102), however, it is [not implemented](https://github.com/decred/dcrd/blob/979d21f6f3695f0081ccc0c55994ecdff39b5a27/rpcserver.go#L302) in the dcrd RPC server and the current results don't contain all of the information pertinent to Decred.

The result struct should be modified to add the following fields:
- `initialblockdownload` (`bool`) - Best guess as to whether the initial block download process is underway
- `chainwork` (`string`) - Hexadecimal value that represents the total work in the current best chain
- `maxblocksize` (`uint32`) - The maximum allowed block size

There should probably also be some additional information regarding the state of the various DCPs, but the best way to present that information needs to be thought through.

It should also be noted that the `verificationprogress` field currently has no code to determine it, so a best guess will need to be determined in the RPC server for now.  In the future, this can be significantly improved through modification to the `blockchain` module, but that is beyond the scope of this issue.

Don't forget when working on this the [JSON-RPC API](https://github.com/decred/dcrd/blob/master/docs/json_rpc_api.md) documentation and [RPC server help](https://github.com/decred/dcrd/blob/master/rpcserverhelp.go) needs to be updated accordingly as well as bumping the RPC server version and potentially the `dcrjson` module version depending on what other changes have happened since the last module release.

### PR DESCRIPTION
**Pull request by dnldd**
This resolves #1475.

Thanks to @davecgh for his help & @matheusd for testing/review.

### CODE REVIEW

File: blockmanager.go [hunk 1]
```diff
@@ -352,6 +357,13 @@ func (b *blockManager) resetHeaderState(newestHash *chainhash.Hash, newestHeight
 	}
 }
 
+// SyncHeight returns the current synced block height from peers.
```
Comment (matheusd) [inline blockmanager.go]: I'd reword this slightly, because reading it, it seems that this is the block height the local node *has* synced, while this is meant to be the block height the local node is trying to sync *to* (ie: the "target" sync height). The semantics for this "syncHeight" are also slighly tricky, since it's either the current best height (on a node that has just loaded up and hasn't tried syncing to anything yet or that has lost all peers) _or_ it's the height it's trying to sync to.

This would also be useful information returned in `getblockchaininfo`, since this rpc already seems to include info that is not strictly from the node's chain.

**Comment by dnldd** [comment]
The value for `since` only changes when the deployment's agenda changes state. Thanks for the review, will run some additional tests.

**Comment by dnldd** [comment]
@matheusd issues raised should be resolved. Here's some test output on mainnet:
```
{
  "chain": "mainnet",
  "blocks": 281147,
  "headers": 281147,
  "syncheight": 281147,
  "bestblockhash": "00000000000000006d29375a4c478dbd7e1f421ce1f3f8f198d6f706ecace041",
  "difficulty": 419468269,
  "verificationprogress": 1,
  "chainwork": "0x92bd67183c9f59c79d50",
  "initialblockdownload": false,
  "maxblocksize": 393216,
  "deployments": {
    "lnfeatures": {
      "version": 5,
      "status": "active",
      "since": 181504,
      "starttime": 1505260800,
      "expiretime": 1536796800
    },
    "lnsupport": {
      "version": 4,
      "status": "active",
      "since": 141184,
      "starttime": 1493164800,
      "expiretime": 1508976000
    },
    "sdiffalgorithm": {
      "version": 4,
      "status": "active",
      "since": 141184,
      "starttime": 1493164800,
      "expiretime": 1524700800
    }
  }
}
```

**Comment by davecgh** [comment]
Those since values do not appear to be correct.

DCP0001 became active at height 149248
DCP0002 became active at height 189568
DCP0003 became active at height 189568

File: blockchain/thresholdstate.go [hunk 2]
```diff
@@ -465,6 +465,82 @@ func (b *BlockChain) deploymentState(prevNode *blockNode, version uint32, deploy
...
+	return b.bestChain.Genesis(), nil
+}
+
+// StateActivationHeight returns the height of the first block where the
+// current state of consensus deployment agenda activated.
+func (b *BlockChain) StateActivationHeight(hash *chainhash.Hash, version uint32, deploymentID string) (int64, error) {
+	// Fetch the block node of the provided hash.
+	node := b.index.LookupNode(hash)
+	if node == nil || !b.index.NodeStatus(node).KnownValid() {
+		return 0, HashError(hash.String())
```
Comment (davecgh) [inline blockchain/thresholdstate.go]: Not the correct error.  It should be the same as the others which involve lookups.  Also, the node status is irrelevant here.

File: blockchain/thresholdstate.go [hunk 3]
```diff
@@ -465,6 +465,82 @@ func (b *BlockChain) deploymentState(prevNode *blockNode, version uint32, deploy
...
+		return 0, fmt.Errorf("Threshold state cache for agenda with "+
+			"deployment id (%s) not found", deploymentID)
+	}
+
+	stateChangeNode, err := b.thresholdStateLastChanged(version, node, checker, cache)
+	if err != nil {
+		return 0, err
+	}
+
+	return stateChangeNode.height + 1, err
```
Comment (davecgh) [inline blockchain/thresholdstate.go]: This is incorrect.  The function which determines the node in question should be returning the correct node.  This would return 2 in the case the genesis block node was returning which is clearly not correct!

File: dcrjson/chainsvrresults.go [hunk 4]
```diff
@@ -89,16 +89,29 @@ type GetBlockVerboseResult struct {
 	NextHash      string        `json:"nextblockhash,omitempty"`
 }
 
+// AgendaInfo provides an overview of an agenda in a consensus deployment.
+type AgendaInfo struct {
+	Version    uint32 `json:"version"`
```
[the remark this answers is not in the document]
Comment (dnldd) [reply dcrjson/chainsvrresults.go]: noted, updated. I initially included it for ease of use with `getvoteinfo`, that requires the development id and the version. `getblockchaininfo` would be requested first and with that the caller would have all the info needed to take a closer look at each vote since the response has the deployment id and the version.

File: rpcserverhelp.go [hunk 5]
```diff
@@ -236,6 +236,25 @@ var helpDescsEnUS = map[string]string{
 	"getblock--condition1": "verbose=true",
 	"getblock--result0":    "Hex-encoded bytes of the serialized block",
 
+	// GetBlockchainInfoCmd help.
+	"getblockchaininfo--synopsis": "Returns information about the current state of the block chain.",
+
+	// GetBlockchainInfoResult help.
+	"getblockchaininforesult-chain":                "The current network name.",
+	"getblockchaininforesult-blocks":               "The block count.",
```
Comment (davecgh) [inline rpcserverhelp.go]: I'd suggest something  a little more descriptive.

Maybe:
```The number of blocks in the current best chain.```

File: rpcserverhelp.go [hunk 6]
```diff
@@ -236,6 +236,25 @@ var helpDescsEnUS = map[string]string{
...
+	"getblockchaininforesult-headers":              "The block header count.",
```
Comment (davecgh) [inline rpcserverhelp.go]: Perhaps:

```The number of validated block headers that comprise the target best chain.```

I'm making this distinction because, while it's not currently the case, the plan is to move to a full headers-first download mode where all of the headers will be downloaded first followed by downloading all of the actual block data in parallel.

File: rpcserverhelp.go [hunk 7]
```diff
@@ -236,6 +236,25 @@ var helpDescsEnUS = map[string]string{
...
+	"getblockchaininforesult-syncheight":           "The latest known block height being synced to.",
+	"getblockchaininforesult-bestblockhash":        "The block hash of the current best chain tip.",
+	"getblockchaininforesult-difficulty":           "The current network difficulty.",
+	"getblockchaininforesult-verificationprogress": "The chain verification progress estimate.",
+	"getblockchaininforesult-chainwork":            "Hex encoded total work done for the chain.",
+	"getblockchaininforesult-initialblockdownload": "Syncing status",
```
Comment (davecgh) [inline rpcserverhelp.go]: This description doesn't seem to be helpful.  I suggest:

```Best guess of whether this node is in the initial block download mode used to catch up the chain when it is far behind```

File: rpcserverhelp.go [hunk 8]
```diff
@@ -236,6 +236,25 @@ var helpDescsEnUS = map[string]string{
...
+	"getblockchaininforesult-maxblocksize":         "The maximum allowed block size.",
+	"getblockchaininforesult-deployments":          "Network consensus deployments.",
+	"getblockchaininforesult-deployments--desc":    "Nonsensus deployment agendas.",
```
Comment (davecgh) [inline rpcserverhelp.go]: s/Nonsensus/Consensus/

File: blockchain/thresholdstate.go [hunk 9]
```diff
@@ -465,6 +465,91 @@ func (b *BlockChain) deploymentState(prevNode *blockNode, version uint32, deploy
...
+		// confirmation window.
+		prevNode = prevNode.RelativeAncestor(confirmationWindow)
+	}
+
+	return b.bestChain.Genesis(), nil
+}
+
+// StateActivationHeight returns the height of the first block where the
+// current state of consensus deployment agenda activated.
+func (b *BlockChain) StateActivationHeight(hash *chainhash.Hash, version uint32, deploymentID string) (int64, error) {
```
Comment (davecgh) [inline blockchain/thresholdstate.go]: This is wrong too.  It's not returning when it activated, rather the height at which the state last changed.  For example, try calling this with the hash at height 140000.

File: rpcserver.go [hunk 10]
```diff
@@ -1903,6 +1904,89 @@ func handleGetBlock(s *rpcServer, cmd interface{}, closeChan <-chan struct{}) (i
...
+
+			state, err := s.chain.NextThresholdState(&best.PrevHash, version,
+				agenda.Vote.Id)
+			if err != nil {
+				return nil, rpcInternalError(err.Error(),
+					fmt.Sprintf("Could not fetch the next threshold state "+
+						"for agenda with id (%v).", agenda.Vote.Id))
+			}
+
+			if state.State != blockchain.ThresholdDefined {
```
Comment (davecgh) [inline rpcserver.go]: No reason to get the state and avoid calling `StateLastChangedHeight` if the state is defined.  The function will just return 0 in that case, which is the same result.

**Comment by davecgh** [comment]
Some results during initial sync with the latest:

```
{
 "chain": "mainnet",
 "blocks": 65239,
 "headers": 65239,
 "syncheight": 281735,
 "bestblockhash": "0000000000000d6f6f779e83c7c78837a1bc206ec27a0b95beeff06d8887d95d",
 "difficulty": 437488364,
 "verificationprogress": 0.23156157381936926,
 "chainwork": "000000000000000000000000000000000000000000000009337ed0b926ca8a88",
 "initialblockdownload": true,
 "maxblocksize": 393216,
 "deployments": {
   "lnfeatures": {
     "status": "defined",
     "starttime": 1505260800,
     "expiretime": 1536796800
   },
   "lnsupport": {
     "status": "defined",
     "starttime": 1493164800,
     "expiretime": 1508976000
   },
   "sdiffalgorithm": {
     "status": "defined",
     "starttime": 1493164800,
     "expiretime": 1524700800
   }
 }
}
{
  "chain": "mainnet",
  "blocks": 133190,
  "headers": 133190,
  "syncheight": 281735,
  "bestblockhash": "0000000000000127851fa41bf9d21d9cd4e6e3d8ccd961e4e503639ddf1f8032",
  "difficulty": 436334663,
  "verificationprogress": 0.472749214687561,
  "chainwork": "00000000000000000000000000000000000000000000002ae7deab0edce62a10",
  "initialblockdownload": true,
  "maxblocksize": 393216,
  "deployments": {
    "lnfeatures": {
      "status": "defined",
      "starttime": 1505260800,
      "expiretime": 1536796800
    },
    "lnsupport": {
      "status": "started",
      "since": 133120,
      "starttime": 1493164800,
      "expiretime": 1508976000
    },
    "sdiffalgorithm": {
      "status": "started",
      "since": 133120,
      "starttime": 1493164800,
      "expiretime": 1524700800
    }
  }
}
{
  "chain": "mainnet",
  "blocks": 141226,
  "headers": 141226,
  "syncheight": 281735,
  "bestblockhash": "00000000000000a4d4d2b8c62d5a605eda0f276cebbf0f24159e633575f78bae",
  "difficulty": 436343246,
  "verificationprogress": 0.5012724723587768,
  "chainwork": "000000000000000000000000000000000000000000000038adb693d55e78938e",
  "initialblockdownload": true,
  "maxblocksize": 393216,
  "deployments": {
    "lnfeatures": {
      "status": "defined",
      "starttime": 1505260800,
      "expiretime": 1536796800
    },
    "lnsupport": {
      "status": "lockedin",
      "since": 141184,
      "starttime": 1493164800,
      "expiretime": 1508976000
    },
    "sdiffalgorithm": {
      "status": "lockedin",
      "since": 141184,
      "starttime": 1493164800,
      "expiretime": 1524700800
    }
  }
}
{
  "chain": "mainnet",
  "blocks": 149344,
  "headers": 149344,
  "syncheight": 281735,
  "bestblockhash": "00000000000000258cc91b5593651d7b31afd5b2f6902cdbd0ffbb3caa65ca85",
  "difficulty": 436309838,
  "verificationprogress": 0.530086783679699,
  "chainwork": "00000000000000000000000000000000000000000000004a2e29de12fc2fb648",
  "initialblockdownload": true,
  "maxblocksize": 393216,
  "deployments": {
    "lnfeatures": {
      "status": "defined",
      "starttime": 1505260800,
      "expiretime": 1536796800
    },
    "lnsupport": {
      "status": "active",
      "since": 149248,
      "starttime": 1493164800,
      "expiretime": 1508976000
    },
    "sdiffalgorithm": {
      "status": "active",
      "since": 149248,
      "starttime": 1493164800,
      "expiretime": 1524700800
    }
  }
}
```
